MacBook M1 & Docker compile error

👋🏻 I'm trying to install/compile Artichoke Ruby in docker on my M1 MacBook but I'm running into a couple of compile errors. Unfortunately, I know next to nothing about Rust, but I will share as much info as I can.

### Dockerfile

```
# syntax=docker/dockerfile:1
FROM rust as builder

WORKDIR /opt

RUN apt update -qq && apt install -y -qq ruby clang && \
  cargo install --git https://github.com/artichoke/artichoke \
  --branch trunk --locked artichoke
```

### Build Error

```
error[E0308]: mismatched types
  --> artichoke-backend/src/extn/core/string/ffi.rs:56:31
   |
56 |     let cstr = CStr::from_ptr(p);
   |                               ^ expected `u8`, found `i8`
   |
   = note: expected raw pointer `*const u8`
              found raw pointer `*const i8`

error[E0308]: mismatched types
  --> artichoke-backend/src/extn/core/string/mod.rs:77:49
   |
77 |                 sys::mrb_sys_alloc_rstring(mrb, ptr.cast::<i8>(), length as sys::mrb_int, capacity as sys::mrb_int)
   |                                                 ^^^^^^^^^^^^^^^^ expected `u8`, found `i8`
   |
   = note: expected raw pointer `*mut u8`
              found raw pointer `*mut i8`

error[E0308]: mismatched types
   --> artichoke-backend/src/extn/core/string/mod.rs:104:17
    |
104 |                 ptr.cast::<i8>(),
    |                 ^^^^^^^^^^^^^^^^ expected `u8`, found `i8`
    |
    = note: expected raw pointer `*mut u8`
               found raw pointer `*mut i8`

error[E0308]: mismatched types
  --> artichoke-backend/src/extn/core/symbol/ffi.rs:38:33
   |
38 |     let string = CStr::from_ptr(name);
   |                                 ^^^^ expected `u8`, found `i8`
   |
   = note: expected raw pointer `*const u8`
              found raw pointer `*const i8`

error[E0308]: mismatched types
  --> artichoke-backend/src/extn/core/symbol/ffi.rs:83:33
   |
83 |     let string = CStr::from_ptr(name);
   |                                 ^^^^ expected `u8`, found `i8`
   |
   = note: expected raw pointer `*const u8`
              found raw pointer `*const i8`

error[E0308]: mismatched types
   --> artichoke-backend/src/extn/core/symbol/ffi.rs:134:33
    |
134 |     let string = CStr::from_ptr(name);
    |                                 ^^^^ expected `u8`, found `i8`
    |
    = note: expected raw pointer `*const u8`
               found raw pointer `*const i8`

error[E0308]: mismatched types
   --> artichoke-backend/src/sys/protect.rs:114:40
    |
114 |         sys::mrb_load_nstring_cxt(mrb, code.as_ptr().cast::<i8>(), code.len(), context)
    |                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `u8`, found `i8`
    |
    = note: expected raw pointer `*const u8`
               found raw pointer `*const i8`

For more information about this error, try `rustc --explain E0308`.
error: failed to compile `artichoke v0.1.0-pre.0 (https://github.com/artichoke/artichoke?branch=trunk#f5a570da)`, intermediate artifacts can be found at `/tmp/cargo-installJutlZX`

Caused by:
  could not compile `artichoke-backend` due to 7 previous errors
```

Almost, now it throws:

```
#9 89.55 error[E0308]: mismatched types
#9 89.55    --> src/parser.rs:137:20
#9 89.55     |
#9 89.55 137 |         parser.s = ptr;
#9 89.55     |                    ^^^ expected `u8`, found `i8`
#9 89.55     |
#9 89.55     = note: expected raw pointer `*const u8`
#9 89.55                found raw pointer `*const i8`
#9 89.55 
#9 89.60 error[E0308]: mismatched types
#9 89.60    --> src/parser.rs:138:32
#9 89.60     |
#9 89.60 138 |         parser.send = unsafe { ptr.offset(len) };
#9 89.60     |                                ^^^^^^^^^^^^^^^ expected `u8`, found `i8`
#9 89.60     |
#9 89.60     = note: expected raw pointer `*const u8`
#9 89.60                found raw pointer `*const i8`
#9 89.60 
#9 89.64 For more information about this error, try `rustc --explain E0308`.
#9 89.65 error: could not compile `artichoke` due to 2 previous errors
#9 89.65 warning: build failed, waiting for other jobs to finish...
#9 91.12 error: build failed
```

This change is missing:
```diff
diff --git a/src/parser.rs b/src/parser.rs
index ae1cf2af77..be7ce27768 100644
--- a/src/parser.rs
+++ b/src/parser.rs
@@ -4,6 +4,7 @@
 //! enter multiline editing mode.
 
 use std::ffi::CStr;
+use std::os::raw::c_char;
 use std::ptr::NonNull;
 
 use crate::backend::sys;
@@ -133,7 +134,7 @@ impl<'a> Parser<'a> {
         let parser = unsafe { self.parser.as_mut() };
         let context = unsafe { self.context.as_mut() };
 
-        let ptr = code.as_ptr().cast::<i8>();
+        let ptr = code.as_ptr().cast::<c_char>();
         parser.s = ptr;
         parser.send = unsafe { ptr.offset(len) };
         parser.lineno = context.lineno;
```

--- a/artichoke-backend/src/extn/core/string/ffi.rs
+++ b/artichoke-backend/src/extn/core/string/ffi.rs
@@ -6,7 +6,7 @@ use core::slice;
 use core::str;
 use std::ffi::{c_void, CStr};
 use std::io::Write as _;
-use std::os::raw::{c_double, c_int};
+use std::os::raw::{c_char, c_double, c_int};
 
 use artichoke_core::convert::Convert;
 use artichoke_core::hash::Hash as _;
@@ -33,7 +33,7 @@ unsafe extern "C" fn mrb_str_new_capa(mrb: *mut sys::mrb_state, capa: usize) -> 
 
 // MRB_API mrb_value mrb_str_new(mrb_state *mrb, const char *p, size_t len)
 #[no_mangle]
-unsafe extern "C" fn mrb_str_new(mrb: *mut sys::mrb_state, p: *const i8, len: usize) -> sys::mrb_value {
+unsafe extern "C" fn mrb_str_new(mrb: *mut sys::mrb_state, p: *const c_char, len: usize) -> sys::mrb_value {
     unwrap_interpreter!(mrb, to => guard);
     let s = if p.is_null() {
         String::utf8(vec![0; len])
@@ -51,7 +51,7 @@ unsafe extern "C" fn mrb_str_new(mrb: *mut sys::mrb_state, p: *const i8, len: us
 
 // MRB_API mrb_value mrb_str_new_cstr(mrb_state *mrb, const char *p)
 #[no_mangle]
-unsafe extern "C" fn mrb_str_new_cstr(mrb: *mut sys::mrb_state, p: *const i8) -> sys::mrb_value {
+unsafe extern "C" fn mrb_str_new_cstr(mrb: *mut sys::mrb_state, p: *const c_char) -> sys::mrb_value {
     unwrap_interpreter!(mrb, to => guard);
     let cstr = CStr::from_ptr(p);
     let bytes = cstr.to_bytes().to_vec();
@@ -65,7 +65,7 @@ unsafe extern "C" fn mrb_str_new_cstr(mrb: *mut sys::mrb_state, p: *const i8) ->
 
 // MRB_API mrb_value mrb_str_new_static(mrb_state *mrb, const char *p, size_t len)
 #[no_mangle]
-unsafe extern "C" fn mrb_str_new_static(mrb: *mut sys::mrb_state, p: *const i8, len: usize) -> sys::mrb_value {
+unsafe extern "C" fn mrb_str_new_static(mrb: *mut sys::mrb_state, p: *const c_char, len: usize) -> sys::mrb_value {
     // Artichoke doesn't have a static string optimization.
     mrb_str_new(mrb, p, len)
 }
@@ -75,7 +75,7 @@ unsafe extern "C" fn mrb_str_new_static(mrb: *mut sys::mrb_state, p: *const i8, 
 unsafe extern "C" fn mrb_str_index(
     mrb: *mut sys::mrb_state,
     s: sys::mrb_value,
-    sptr: *const i8,
+    sptr: *const c_char,
     slen: sys::mrb_int,
     offset: sys::mrb_int,
 ) -> sys::mrb_int {
@@ -335,12 +335,12 @@ unsafe extern "C" fn mrb_ptr_to_str(mrb: *mut sys::mrb_state, p: *mut c_void) ->
 //
 // obsolete: use RSTRING_CSTR() or mrb_string_cstr()
 #[no_mangle]
-unsafe extern "C" fn mrb_string_value_cstr(mrb: *mut sys::mrb_state, ptr: *mut sys::mrb_value) -> *const i8 {
+unsafe extern "C" fn mrb_string_value_cstr(mrb: *mut sys::mrb_state, ptr: *mut sys::mrb_value) -> *const c_char {
     unwrap_interpreter!(mrb, to => guard, or_else = ptr::null());
     let mut s = Value::from(*ptr);
     let mut string = if let Ok(string) = String::unbox_from_value(&mut s, &mut guard) {
         if let Some(b'\0') = string.last() {
-            return string.as_ptr().cast::<i8>();
+            return string.as_ptr().cast();
         }
         string
     } else {
@@ -356,7 +356,7 @@ unsafe extern "C" fn mrb_string_value_cstr(mrb: *mut sys::mrb_state, ptr: *mut s
     //
     // This raw pointer will not be invalidated since we rebox this `String`
     // into the mruby heap where the GC will keep it alive.
-    let cstr = string.as_ptr().cast::<i8>();
+    let cstr = string.as_ptr().cast::<c_char>();
 
     let inner = string.take();
     String::box_into_value(inner, s, &mut guard).expect("String reboxing should not fail");
@@ -366,12 +366,12 @@ unsafe extern "C" fn mrb_string_value_cstr(mrb: *mut sys::mrb_state, ptr: *mut s
 
 // MRB_API const char* mrb_string_cstr(mrb_state *mrb, mrb_value str)
 #[no_mangle]
-unsafe extern "C" fn mrb_string_cstr(mrb: *mut sys::mrb_state, s: sys::mrb_value) -> *const i8 {
+unsafe extern "C" fn mrb_string_cstr(mrb: *mut sys::mrb_state, s: sys::mrb_value) -> *const c_char {
     unwrap_interpreter!(mrb, to => guard, or_else = ptr::null());
     let mut s = Value::from(s);
     let mut string = if let Ok(string) = String::unbox_from_value(&mut s, &mut guard) {
         if let Some(b'\0') = string.last() {
-            return string.as_ptr().cast::<i8>();
+            return string.as_ptr().cast();
         }
         string
     } else {
@@ -387,7 +387,7 @@ unsafe extern "C" fn mrb_string_cstr(mrb: *mut sys::mrb_state, s: sys::mrb_value
     //
     // This raw pointer will not be invalidated since we rebox this `String`
     // into the mruby heap where the GC will keep it alive.
-    let cstr = string.as_ptr().cast::<i8>();
+    let cstr = string.as_ptr().cast::<c_char>();
 
     let inner = string.take();
     String::box_into_value(inner, s, &mut guard).expect("String reboxing should not fail");
@@ -499,7 +499,7 @@ unsafe extern "C" fn mrb_str_to_dbl(mrb: *mut sys::mrb_state, s: sys::mrb_value,
 unsafe extern "C" fn mrb_str_cat(
     mrb: *mut sys::mrb_state,
     s: sys::mrb_value,
-    ptr: *const i8,
+    ptr: *const c_char,
     len: usize,
 ) -> sys::mrb_value {
     unwrap_interpreter!(mrb, to => guard, or_else = s);

--- a/artichoke-backend/src/extn/core/string/mod.rs
+++ b/artichoke-backend/src/extn/core/string/mod.rs
@@ -1,5 +1,6 @@
 use core::ops::Deref;
 use std::ffi::c_void;
+use std::os::raw::c_char;
 
 use artichoke_core::value::Value as _;
 use spinoso_exception::TypeError;
@@ -74,7 +75,12 @@ impl BoxUnboxVmValue for String {
         let RawParts { ptr, length, capacity } = String::into_raw_parts(value);
         let value = unsafe {
             interp.with_ffi_boundary(|mrb| {
-                sys::mrb_sys_alloc_rstring(mrb, ptr.cast::<i8>(), length as sys::mrb_int, capacity as sys::mrb_int)
+                sys::mrb_sys_alloc_rstring(
+                    mrb,
+                    ptr.cast::<c_char>(),
+                    length as sys::mrb_int,
+                    capacity as sys::mrb_int,
+                )
             })?
         };
         let string = unsafe { sys::mrb_sys_basic_ptr(value).cast::<sys::RString>() };
@@ -101,7 +107,7 @@ impl BoxUnboxVmValue for String {
         let RawParts { ptr, length, capacity } = String::into_raw_parts(value);
         let string = unsafe {
             sys::mrb_sys_repack_into_rstring(
-                ptr.cast::<i8>(),
+                ptr.cast::<c_char>(),
                 length as sys::mrb_int,
                 capacity as sys::mrb_int,
                 into.inner(),

--- a/artichoke-backend/src/extn/core/symbol/ffi.rs
+++ b/artichoke-backend/src/extn/core/symbol/ffi.rs
@@ -1,4 +1,5 @@
 use std::ffi::CStr;
+use std::os::raw::c_char;
 use std::ptr;
 use std::slice;
 
@@ -9,7 +10,7 @@ use crate::extn::prelude::*;
 // MRB_API mrb_sym mrb_intern(mrb_state*,const char*,size_t);
 //
 #[no_mangle]
-unsafe extern "C" fn mrb_intern(mrb: *mut sys::mrb_state, name: *const i8, len: usize) -> sys::mrb_sym {
+unsafe extern "C" fn mrb_intern(mrb: *mut sys::mrb_state, name: *const c_char, len: usize) -> sys::mrb_sym {
     let bytes = slice::from_raw_parts(name.cast::<u8>(), len);
     let bytes = bytes.to_vec();
     unwrap_interpreter!(mrb, to => guard, or_else = 0);
@@ -22,7 +23,7 @@ unsafe extern "C" fn mrb_intern(mrb: *mut sys::mrb_state, name: *const i8, len: 
 // MRB_API mrb_sym mrb_intern_static(mrb_state*,const char*,size_t);
 // ```
 #[no_mangle]
-unsafe extern "C" fn mrb_intern_static(mrb: *mut sys::mrb_state, name: *const i8, len: usize) -> sys::mrb_sym {
+unsafe extern "C" fn mrb_intern_static(mrb: *mut sys::mrb_state, name: *const c_char, len: usize) -> sys::mrb_sym {
     let bytes = slice::from_raw_parts::<'static, _>(name.cast::<u8>(), len);
     unwrap_interpreter!(mrb, to => guard, or_else = 0);
     let sym = guard.intern_bytes(bytes);
@@ -34,7 +35,7 @@ unsafe extern "C" fn mrb_intern_static(mrb: *mut sys::mrb_state, name: *const i8
 // MRB_API mrb_sym mrb_intern_cstr(mrb_state *mrb, const char* str);
 // ```
 #[no_mangle]
-unsafe extern "C" fn mrb_intern_cstr(mrb: *mut sys::mrb_state, name: *const i8) -> sys::mrb_sym {
+unsafe extern "C" fn mrb_intern_cstr(mrb: *mut sys::mrb_state, name: *const c_char) -> sys::mrb_sym {
     let string = CStr::from_ptr(name);
     let bytes = string.to_bytes_with_nul().to_vec();
     unwrap_interpreter!(mrb, to => guard, or_else = 0);
@@ -65,7 +66,7 @@ unsafe extern "C" fn mrb_intern_str(mrb: *mut sys::mrb_state, name: sys::mrb_val
 // MRB_API mrb_sym mrb_intern_check(mrb_state*,const char*,size_t);
 // ```
 #[no_mangle]
-unsafe extern "C" fn mrb_intern_check(mrb: *mut sys::mrb_state, name: *const i8, len: usize) -> sys::mrb_sym {
+unsafe extern "C" fn mrb_intern_check(mrb: *mut sys::mrb_state, name: *const c_char, len: usize) -> sys::mrb_sym {
     let bytes = slice::from_raw_parts(name.cast::<u8>(), len);
     unwrap_interpreter!(mrb, to => guard, or_else = 0);
     if let Ok(Some(sym)) = guard.check_interned_bytes(bytes) {
@@ -79,7 +80,7 @@ unsafe extern "C" fn mrb_intern_check(mrb: *mut sys::mrb_state, name: *const i8,
 // MRB_API mrb_sym mrb_intern_check_cstr(mrb_state*,const char*);
 // ```
 #[no_mangle]
-unsafe extern "C" fn mrb_intern_check_cstr(mrb: *mut sys::mrb_state, name: *const i8) -> sys::mrb_sym {
+unsafe extern "C" fn mrb_intern_check_cstr(mrb: *mut sys::mrb_state, name: *const c_char) -> sys::mrb_sym {
     let string = CStr::from_ptr(name);
     let bytes = string.to_bytes();
     unwrap_interpreter!(mrb, to => guard, or_else = 0);
@@ -115,7 +116,7 @@ unsafe extern "C" fn mrb_intern_check_str(mrb: *mut sys::mrb_state, name: sys::m
 // MRB_API mrb_value mrb_check_intern(mrb_state*,const char*,size_t);
 // ```
 #[no_mangle]
-unsafe extern "C" fn mrb_check_intern(mrb: *mut sys::mrb_state, name: *const i8, len: usize) -> sys::mrb_value {
+unsafe extern "C" fn mrb_check_intern(mrb: *mut sys::mrb_state, name: *const c_char, len: usize) -> sys::mrb_value {
     let bytes = slice::from_raw_parts(name.cast::<u8>(), len);
     unwrap_interpreter!(mrb, to => guard);
     let symbol = if let Ok(Some(sym)) = guard.check_interned_bytes(bytes) {
@@ -130,7 +131,7 @@ unsafe extern "C" fn mrb_check_intern(mrb: *mut sys::mrb_state, name: *const i8,
 // MRB_API mrb_value mrb_check_intern_cstr(mrb_state*,const char*);
 // ```
 #[no_mangle]
-unsafe extern "C" fn mrb_check_intern_cstr(mrb: *mut sys::mrb_state, name: *const i8) -> sys::mrb_value {
+unsafe extern "C" fn mrb_check_intern_cstr(mrb: *mut sys::mrb_state, name: *const c_char) -> sys::mrb_value {
     let string = CStr::from_ptr(name);
     let bytes = string.to_bytes_with_nul();
     unwrap_interpreter!(mrb, to => guard);
@@ -182,7 +183,7 @@ unsafe extern "C" fn mrb_sym_name_len(
     mrb: *mut sys::mrb_state,
     sym: sys::mrb_sym,
     lenp: *mut sys::mrb_int,
-) -> *const i8 {
+) -> *const c_char {
     if !lenp.is_null() {
         ptr::write(lenp, 0);
     }
@@ -195,7 +196,7 @@ unsafe extern "C" fn mrb_sym_name_len(
                 return ptr::null();
             }
         }
-        bytes.as_ptr().cast::<i8>()
+        bytes.as_ptr().cast()
     } else {
         ptr::null()
     }
@@ -205,7 +206,7 @@ unsafe extern "C" fn mrb_sym_name_len(
 // MRB_API const char *mrb_sym_dump(mrb_state*,mrb_sym);
 // ```
 #[no_mangle]
-unsafe extern "C" fn mrb_sym_dump(mrb: *mut sys::mrb_state, sym: sys::mrb_sym) -> *const i8 {
+unsafe extern "C" fn mrb_sym_dump(mrb: *mut sys::mrb_state, sym: sys::mrb_sym) -> *const c_char {
     unwrap_interpreter!(mrb, to => guard, or_else = ptr::null());
     if let Ok(Some(bytes)) = guard.lookup_symbol(sym) {
         let bytes = bytes.to_vec();
@@ -213,7 +214,7 @@ unsafe extern "C" fn mrb_sym_dump(mrb: *mut sys::mrb_state, sym: sys::mrb_sym) -
         // a pointer to it.
         if let Ok(string) = guard.try_convert_mut(bytes) {
             if let Ok(bytes) = string.try_convert_into_mut::<&[u8]>(&mut guard) {
-                return bytes.as_ptr().cast::<i8>();
+                return bytes.as_ptr().cast();
             }
         }
     }

--- a/artichoke-backend/src/sys/protect.rs
+++ b/artichoke-backend/src/sys/protect.rs
@@ -1,5 +1,6 @@
 use std::ffi::c_void;
 use std::mem;
+use std::os::raw::c_char;
 use std::ptr::{self, NonNull};
 
 use crate::sys;
@@ -111,7 +112,7 @@ impl<'a> Protect for Eval<'a> {
         // `mrb_load_nstring_ctx` sets the "stack keep" field on the context
         // which means the most recent value returned by eval will always be
         // considered live by the GC.
-        sys::mrb_load_nstring_cxt(mrb, code.as_ptr().cast::<i8>(), code.len(), context)
+        sys::mrb_load_nstring_cxt(mrb, code.as_ptr().cast::<c_char>(), code.len(), context)
     }
 }
 

--- a/src/parser.rs
+++ b/src/parser.rs
@@ -4,6 +4,7 @@
 //! enter multiline editing mode.
 
 use std::ffi::CStr;
+use std::os::raw::c_char;
 use std::ptr::NonNull;
 
 use crate::backend::sys;
@@ -133,7 +134,7 @@ impl<'a> Parser<'a> {
         let parser = unsafe { self.parser.as_mut() };
         let context = unsafe { self.context.as_mut() };
 
-        let ptr = code.as_ptr().cast::<i8>();
+        let ptr = code.as_ptr().cast::<c_char>();
         parser.s = ptr;
         parser.send = unsafe { ptr.offset(len) };
         parser.lineno = context.lineno;

